### ISSUE TITLE
[RFC] GLA Kernel + LightningAttnBackend Integration

### ISSUE THREAD

**Issue by cjx0709** [sgl-project/sglang-jax#1032]
## Motivation

This is subtask 2 of #1027 (Support Ling2.5 Series). After the recurrent state memory pool system (#1034) and `HybridLinearAttnBackend` are merged into main, we need to land the GLA (Gated Linear Attention) kernel integration as `LightningAttnBackend`.

## Prerequisites

- [ ] `RecurrentStatePool` merged to main  (#1034 )
- [ ] `HybridLinearAttnBackend` merged to main

### In scope

1. **Extract `LightningAttnBackend(LinearRecurrentAttnBackend)`** from `BailingMoeV2_5LinearAttention`
   - Decode via `fused_recurrent_simple_gla` (jax.lax.scan)
   - Chunked prefill via `simple_gla_fwd` (Pallas kernel); kernel pads each varlen sequence to `chunk_size` internally, so callers pass tight `[1, T_outer, H, K]` directly — no scatter/gather buffer needed
   - State management through `RecurrentStatePool` (no conv state, `conv_kernel_size=0`)
   - Per-layer ALiBi slope owned by `LightningAttnBackend.tp_slope[layer_id]` (mirrors upstream sglang `LightningAttentionBackend.tp_slope`)
   - Metadata: `LinearRecurrentAttnBackendMetadata` with real (non-aligned) `cu_q_lens`, inherited from base class

2. **Wire into `HybridLinearAttnBackend` dispatcher** (`attn_backend_wrapper` factory recognizes Ling-2.5 config → creates `LightningAttnBackend`)

3. **DP adaptation** (to be implemented during rebase onto main with DP support)
   - `get_forward_metadata()` handles DP-padded batch layout
   - `shard_map` in_specs/out_specs: add `"data"` axis to batch dimension
   - `cu_q_lens` / `recurrent_indices` become per-DP-shard
   - Decode path: `fused_recurrent_simple_gla` batch dimension sharded along `"data"`

4. **Simplify `BailingMoeV2_5LinearAttention`** to pure model layer (proj + norm + gate); attention dispatch goes through a new `RadixLightningAttention` wrapper (mirrors KDA's `RadixLinearAttention`), which forwards to `forward_batch.attn_backend(..., pool=recurrent_state_pool)`. Model layer no longer takes a `backend` constructor arg and no longer owns slope.

5. **Align directory structure with upstream sglang** — `fla/` keeps only low-level kernel/norm ops; backend lives in `linear/lightning_backend.py`.

6. **Model layer dtype alignment with GPU** — `BailingMoeV2_5LinearAttention` casts QKV to `float32` after projection (before QK norm / RoPE / kernel dispatch), and casts attention output back to input dtype (before gating). Matches GPU `bailing_moe_linear.py:530` (`qkv = qkv.to(torch.float32)`) and line 572 (`hidden = self.attn(q, k, v, forward_batch).to(hidden_states.dtype)`).

### Out of scope

- GLA model skeleton integration (Bailing MoE v2.5 modeling)
- Kernel performance tuning (chunk_size, split_k, etc.)

### New files

| File | Description |
|------|-------------|
| `srt/layers/radix_lightning_attention.py` | `RadixLightningAttention` — thin nnx.Module dispatcher; mirrors `RadixLinearAttention` for the KDA family |
| `srt/layers/attention/linear/lightning_backend.py` | `LightningAttnBackend` — decode, extend, state management; owns per-layer `tp_slope` dict |
| `srt/kernels/simple_gla/simple_gla.py` | Vendor Pallas kernel — kernel pads varlen sequences to `chunk_size` internally |
| `test/test_lightning_backend.py` | Backend end-to-end validation (18 tests) — JAX naive jit reference (same dtype), validates backend wiring + kernel output |
| `test/layers/test_lightning_backend_dp.py` | DP correctness tests (7 tests) — DP=2×TP=2 validation |
| `test/layers/test_gla_shard.py` | Shard precision tests (6 tests) — production-like H=16 K=128, numpy float64 reference |
| `test/layers/test_gla.py` | Model-level tests (4 tests) — slope formula, torch HF parity, TP consistency |
| `test/layers/test_gla_backend.py` | Backend unit tests (5 tests) — hybrid routing, factory wire-up, fail-fast, DP sharding contracts |

### Modified files

| File | Change |
|------|--------|
| `srt/layers/attention/linear/__init__.py` | Export `LightningAttnBackend` |
| `srt/layers/attention/hybrid_linear_attn_backend.py` | Add `conv_kernel_size=0` support for GLA (no conv state); `attn_backend_wrapper` instantiates `LightningAttnBackend` with `linear_recurrent_layer_ids / num_hidden_layers / num_heads` for `tp_slope` precomputation |
| `srt/mem_cache/recurrent_state_pool.py` | Allow `conv_kernel_size=0` in pool allocation |
| `models/bailing_moe_v2_5_linear_attention.py` | Simplify to pure model layer; `self.attn = RadixLightningAttention(...)` for dispatch; cast QKV to float32 after projection (matching GPU `bailing_moe_linear.py:530`); cast attention output back to input dtype (matching GPU line 572) |

### DP adaptation changes (to be done during rebase)

| File | Change |
|------|--------|
| `lightning_backend.py` | `_forward_extend` shard_map: `P(None, "tensor", None)` → `P("data", "tensor", None)` for q/k/v; `cu_q_lens` per-DP-shard |
| `lightning_backend.py` | `_forward_decode`: ensure batch dim sharded along `"data"` |
| `test/layers/test_lightning_backend_dp.py` | Add 7 DP tests with `mesh = create_device_mesh(ici_parallelism=[2, 2])` |

# Test Plan

Tests are organized bottom-up by abstraction level. Each test verifies a unique correctness property — no redundancy across files.

## Precision Standard

Tolerance values are aligned with the GPU sglang FLA CI convention defined in [`sglang/python/sglang/srt/layers/attention/fla/utils.py:84-97`](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/attention/fla/utils.py#L84-L97):

```python
def assert_close(prefix, ref, tri, ratio, warning=False, err_atol=1e-6):
    abs_atol = get_abs_err(ref, tri)       # max |ref - tri|
    error_rate = get_err_ratio(ref, tri)    # RMSE / RMS(ref)
    if abs_atol <= err_atol:
        return
    if FLA_CI_ENV and (error_rate < 0.01 or abs_atol <= 0.3):
        warnings.warn(msg)                 # pass with warning
    else:
        assert error_rate < ratio, msg     # hard fail
```

**Key rule**: In GPU FLA CI, a test **passes** (with warning) when `abs_atol <= 0.3` **OR** `error_rate < 1%`. This 0.3 bound is the GPU team's accepted maximum absolute error for bf16 kernel outputs.

**Our tolerance mapping**:

| Dtype | Metric | Tolerance | Rationale |
|-------|--------|-----------|-----------|
| bf16 | output atol | 3e-1 | Matches GPU FLA CI `abs_atol <= 0.3` bound |
| bf16 | output rtol | 2e-2 | Within GPU `error_rate < 0.01` bound |
| bf16 | state atol | 1e-1 | State accumulates in fp32, tighter than output |
| bf16 | state rtol | 2e-2 | Consistent with output rtol |
| fp32 | output atol (extend) | 1e-1 | Chunk kernel vs sequential scan have different accumulation order |
| fp32 | output atol (decode) | 1e-3 | Single-step recurrence is essentially exact |
| fp32 | state atol (extend) | 1e-3 | State stays tight even in chunk kernel |
| fp32 | state atol (decode) | 1e-4 | Single-step state update is essentially exact |

**Why not tighter?** GLA is a linear recurrence that accumulates state without normalization (`h_t = exp(g) * h_{t-1} + k_t ⊗ v_t`). Unlike softmax attention, output magnitude grows with sequence length. The chunk kernel (prefill) partitions the sequence into fixed-size chunks and reassembles — this introduces accumulation order differences vs the sequential scan reference, leading to ~0.1 max absolute error even in fp32. This is an inherent algorithmic difference, not a precision bug.

**Test inputs use unit-scale** (no artificial `* 0.1` scaling) — q, k, v, h0 are drawn from `standard_normal`. This matches production conditions where `BailingMoeV2_5LinearAttention` feeds float32-cast QKV directly into the kernel (GPU `bailing_moe_linear.py:530`).

## 1. Backend unit tests — `test_gla_backend.py` (5 tests)

| Class | Test | What it catches |
|-------|------|-----------------|
| TestHybridIntegration | `test_dispatch_routes_to_lightning` | `HybridLinearAttnBackend` routes non-full-attn layers to `LightningAttnBackend` |
| TestHybridIntegration | `test_attn_backend_wrapper_is_noop_until_model_runner_config_is_wired` | `attn_backend_wrapper` stays conservative until model-runner config is wired |
| TestTpSlopeFailFast | `test_keyerror_message_lists_registered_ids` | Missing `tp_slope[layer_id]` raises a contextual `KeyError` listing registered ids and pointing at `attn_backend_wrapper` |
| TestDPShardingContracts | `test_linear_metadata_dp_uses_data_sharding` | DP metadata is laid out as per-rank sections and sharded on `"data"` |
| TestDPShardingContracts | `test_extend_dp_shard_map_uses_local_token_and_state_specs` | DP extend must shard q/k/v, state buffer, indices, and cu_lens together |

## 2. Model-level tests — `test_gla.py` (4 tests)

| Class | Test | What it catches | Tolerance |
|-------|------|-----------------|-----------|
| TestSlopes | `test_slopes_match_hf_formula` | ALiBi slope values (read from `backend.tp_slope[layer_id]`) match HF formula for layers 1, 10, 40, 79 | exact |
| TestModuleLevelMockKernel | `test_forward_with_mocked_kernel` | Complete model layer pipeline (QKV proj → norm → RoPE → backend → gating → dense) matches torch HF reference | atol=1e-2 rtol=2e-2 |
| TestTPConsistency | `test_decode_tp_matches_tp1` | TP=N decode output/state matches TP=1 | atol=1e-2 rtol=2e-2 |
| TestTPConsistency | `test_prefill_tp_matches_tp1` | TP=N prefill output/state matches TP=1 | atol=1e-2 rtol=2e-2 |

## 3. Backend end-to-end validation — `test/test_lightning_backend.py` (18 tests)

**Reference**: JAX naive jit implementation (same dtype as kernel) using per-request scan + jnp.einsum, no Pallas. Validates backend wiring (pool ↔ kernel, metadata ↔ kernel, shard_map ↔ kernel, cross-call state continuity, slope indexing) combined with kernel output end-to-end.

### 3.1 Prefill path (Pallas chunk kernel `chunk_simple_gla_fwd_varlen`) — 9 tests

| Test | Dimension | Config | Tolerance (out / state) | What it catches |
|------|-----------|--------|-------------------------|-----------------|
| `test_extend_single_aligned` | Single-request baseline | `lens=[64]` H=4 K=128 bf16 | 3e-1 / 1e-1 | Basic e2e flow (QKV → backend → pool → kernel → output) |
| `test_extend_two_aligned` | Multi-request cu_seqlens boundary | `lens=[64, 128]` | 3e-1 / 1e-1 | Backend metadata `cu_q_lens` correctly wired to kernel |
| `test_extend_three_aligned_mixed` | Multi-request length diversity | `lens=[64, 192, 128]` | 3e-1 / 1e-1 | Metadata + recurrent_indices multi-case flow |
| `test_extend_with_trailing_empty_slots` | DP padding empty slots | `lens=[64, 128]`+`n_padded=8` | 3e-1 / 1e-1 | Backend `set_ssm_state` does not write to wrong positions / pollute pool |
| `test_extend_with_nonzero_initial_state` | h0 flow from pool | `lens=[128]` + nonzero h0 | 3e-1 / 1e-1 | Backend `get_state(pool)` → kernel `h0` flow |
| `test_extend_h64_k128_full_ling25` | Production sharding | H=64 K=128 full Ling-2.5 heads | 3e-1 / 1e-1 | `shard_map` behavior correct at production shape |
| `test_extend_fp32_strict` | fp32 numerical path | `lens=[128]` fp32 | 1e-1 / 1e-3 | Backend complete fp32 path (pool fp32 buffer / reshard / shard_map output) |
| `test_extend_layer_id_varies` | Slope indexing | 3 different `layer_id` on same input, outputs mutually unequal | — | `backend.tp_slope[layer_id]` indexing correctly passed to kernel |
| `test_extend_then_decode_long_state_continuity` | Cross-path state continuity | 4096-token extend → 32 decode steps | 3e-1 / 1e-1 | Pool state continuity across prefill/decode paths |

### 3.2 Decode path (pure JAX `fused_recurrent_simple_gla`) — 9 tests

| Test | Dimension | Config | Tolerance (out / state) | What it catches |
|------|-----------|--------|-------------------------|-----------------|
| `test_decode_single_request` | Single-request baseline | batch=1 single-step | 3e-1 / 1e-1 | Basic e2e flow |
| `test_decode_batch_4` | Multi-request state isolation | batch=4 | 3e-1 / 1e-1 | Backend multi-request pool slot indexing isolation |
| `test_decode_with_trailing_empty_slots` | DP padding | batch=2+`n_padded=8` | 3e-1 / 1e-1 | `set_ssm_state` does not pollute empty slots |
| `test_decode_state_propagates_3_steps` | Pool RW chain | batch=2 across 3 steps | — | Pool read-modify-write chain correct (state evolves from zero) |
| `test_decode_state_propagates_long_32_steps` | Long-range pool RW | batch=2 across 32 steps | 3e-1 / 1e-1 | Pool repeated RW does not degrade |
| `test_decode_with_nonzero_initial_state` | h0 flow from pool | batch=2 + nonzero h0 single-step | 3e-1 / 1e-1 | `get_state(pool)` → kernel `h0` flow |
| `test_decode_layer_id_boundary` | Boundary layer_id slope | `layer_id ∈ {0, num_hidden_layers-1}` | 1e-2 / 1e-2 | Boundary layer_id slope indexing |
| `test_decode_h64_k128_full_ling25` | Production sharding | H=64 K=128 full heads | 3e-1 / 1e-1 | `shard_map` decode path behavior at production shape |
| `test_decode_fp32_strict` | fp32 numerical path | fp32 | 1e-3 / 1e-4 | Backend complete fp32 path |

## 4. Shard precision tests — `test/layers/test_gla_shard.py` (6 tests)

Production-like dimensions (H=16, K=128, TP=4 → H_local=4). Reference: pure-numpy float64 GLA recurrence.

| Class | Test | What it catches | Tolerance |
|-------|------|-----------------|-----------|
| TestDecodeSharded | `test_decode_single_request` | Kernel vs numpy reference at production H_local | atol=5e-2 rtol=2e-2 |
| TestDecodeSharded | `test_decode_multi_step` | 32-step autoregressive decode, B=4 — state drift | atol=1e-1 rtol=2e-2 |
| TestExtendSharded | `test_extend_single_short` | Extend short sequence (100 tokens) | atol=3e-1 rtol=2e-2 |
| TestExtendSharded | `test_extend_multi_request_variable` | Extend [256, 100, 512] — variable lengths | atol=3e-1 rtol=2e-2 |
| TestExtendSharded | `test_extend_with_nonzero_initial_state` | Extend 4096 tokens with nonzero h0 | atol=3e-1 rtol=2e-2 |
| TestEndToEndBackend | `test_extend_then_decode` | Extend 4096 → 32 decode steps — state continuity | atol=5e-2 |

## 5. DP correctness — `test/layers/test_lightning_backend_dp.py` (7 tests)

DP=2, TP=2 on TPU v6e-4 with `mesh = create_device_mesh(ici_parallelism=[2, 2])`. Ground truth: DP=1×TP=4 reference.

| Class | Test | What it catches | Tolerance |
|-------|------|-----------------|-----------|
| TestDPMetadata | `test_dp_metadata_per_shard` | Each DP shard's `cu_q_lens` independently correct | exact |
| TestDPDecode | `test_decode_dp2_tp2_matches_tp4` | DP=2×TP=2 decode output/state matches DP=1×TP=4 | atol=1e-3 |
| TestDPDecode | `test_decode_dp_state_isolation` | No cross-DP state leakage (distinct h0 per shard) | atol=1e-3 |
| TestDPExtend | `test_extend_dp2_tp2_matches_tp4` | DP=2×TP=2 extend output/state matches DP=1×TP=4 | atol=1e-2 |
| TestDPExtend | `test_extend_dp_multi_request` | Multi-request extend under DP | atol=1e-2 |
| TestDPEndToEnd | `test_dp_extend_then_decode` | Full extend→decode flow under DP | atol=1e-2 |
| TestDPEndToEnd | `test_dp_pool_state_roundtrip` | Each DP shard's pool state roundtrip | atol=1e-3 |

**Total: 40 tests** (5 unit + 4 model + 18 backend e2e + 6 shard + 7 DP). Pass criteria: 40/40.

## Known Constraints

- **Pool state dtype**: `RecurrentStatePool` must use `float32` for GLA state buffers. Using `bfloat16` causes silent truncation when the kernel's `float32` output state is written back.
- **Model layer dtype**: `BailingMoeV2_5LinearAttention` must cast QKV to `float32` after projection (before QK norm / RoPE / kernel), matching GPU `bailing_moe_linear.py:530`. Without this cast, q/k/v enter the kernel in bf16 instead of float32, causing precision degradation that is masked by artificial input scaling in tests but impacts end-to-end model accuracy.

## References

- Parent issue: #1027
- Dependency: #1034 (Recurrent state pool)
- GPU precision standard: [`sglang/python/sglang/srt/layers/attention/fla/utils.py:84-97`](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/attention/fla/utils.py#L84-L97)
- GPU model layer dtype: [`sglang/python/sglang/srt/models/bailing_moe_linear.py:530,572`](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/models/bailing_moe_linear.py)
